runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,219 @@
+        if (subscriptions.size() >= MAX_SUBSCRIBERS) {
+            throw new SubscriberAlreadyExistsException(
+                    "subscribe: Max "+MAX_SUBSCRIBERS+" hit! Consider consolidating "+streamListener);
+        }
+        if (subscriptions.containsKey(streamListener)) {
+            throw new SubscriberAlreadyExistsException(
+                    "StreamManager:subscrbie already registered "+streamListener+". Maybe call unregister");
+        }
+        log.info("StreamManager::subscribe {}, startAddress {}, namespace {}, tables {}", streamListener,
+                startAddress, namespace, tablesOfInterest.toString());
+        SubscriptionThread st = new SubscriptionThread(this, streamListener,
+                namespace, tablesOfInterest, startAddress);
+        subscriptions.put(streamListener, st);
+        st.start();
+    }
+
+    /**
+     * Unsubscribe a prior subscription.
+     *
+     * @param streamListener Client listener.
+     */
+    public synchronized void unsubscribe(@Nonnull StreamListener streamListener) {
+        SubscriptionThread st = subscriptions.remove(streamListener);
+        if (st == null) {
+            log.warn("StreamManager::unsubscribe has no context for "+streamListener.toString());
+            return;
+        }
+        st.setShutdown(true);
+        st.interrupt();
+        log.info("Unsubscribed StreamListener {}", streamListener.toString());
+    }
+
+    class SubscriptionThread extends Thread {
+        /**
+         * The callback context
+         */
+        private final StreamListener listener;
+
+        /**
+         * Parent class to manage errors.
+         */
+        @Getter
+        private final StreamManager streamManager;
+
+        /**
+         * Namespace of interest.
+         */
+        @Getter
+        private final String namespace;
+
+        /**
+         * Tables of interest.
+         *
+         * Map of (Stream Id - TableSchema)
+         */
+        @Getter
+        private final Map<UUID,
+                TableSchema<? extends Message, ? extends Message, ? extends Message>> tablesOfInterest;
+
+        /**
+         * Starting address.
+         */
+        private final long startAddress;
+
+        @Getter
+        @Setter
+        private volatile boolean shutdown = false;
+
+        /**
+         * A reference to the underlying stream
+         */
+        @Getter
+        private final IStreamView txnStream;
+
+        public void run() {
+            try {
+                log.info("{}:Seeking txStream to {}", listener.toString(), startAddress + 1);
+                txnStream.seek(startAddress + 1);
+            } catch (Throwable throwable) {
+                log.warn("{}:onError at seek to {}", listener.toString(), startAddress + 1);
+                listener.onError(throwable);
+                streamManager.unsubscribe(listener);
+                return;
+            }
+
+            while (!isShutdown()) {
+                try {
```

Comment:
I think it's cleaner to put the try catch block outside the while loop, because once it hits exception, it is stopped. And preferrably also combine the the first try catch block for txnStream.seek(), they seem to be the same.

Reply:
I will put the try catch outside the while, but can't combine it with the first.
The first exception is distinctly different. It happens when a subscriber is trying to start from a trimmed address which is different from a subscriber hitting a trimmed exception during normal operation.

Reply:
Actually seek() should not raise any exception as it just moves the pointer, but it's ok to be defensive.